Feature/viewer title

interim17 opened the pull request:
Closes #9 

Changes of note:
- In `widget.ts` we parse the trajectory info and pass `modelInfo` -> `Viewer` -> `ViewerTitle`
- `ViewerTitle` renders the title
-  `viewer.css` centers the viewer under the title
- Styling deviates slightly from design brief on some details: I kept font size at 20px but it looks a bit big to me, dropped the font weight to 400 because 600 looked way too bold to my eye.
- Icon `hover` and `click` styling is in place but info panel pop-up will be in a later PR, I find the hover color from the designs (#E7E4F2) so subtle as to be almost unnoticeable, @TravisKroeker thoughts on that? Screenshots below.
- `ViewerTitle` checks for any metadata other than the title and renders the icon if that's true
- Add a `constants.css` file for font import and global style rule
- Add a `constants.ts` file for `ModelInfo` interface
- Removed placeholder playback controls: not needed and messing with layout, actual playback controls forthcoming
- Colors are being assigned with direct hexes, pending approval of [this PR](https://github.com/simularium/nbsv/pull/8)

Screenshots:

Whole widget:
![Screenshot 2023-11-28 at 6 03 33 PM](https://github.com/simularium/nbsv/assets/24981838/892d1e30-19c7-44c3-a7d8-5a22fa0957b1)

Icon neutral:
![Screenshot 2023-11-28 at 4 54 19 PM](https://github.com/simularium/nbsv/assets/24981838/69d25f6d-bb03-468d-9f4c-3b3bb1bb3190)

Icon hover:
![Screenshot 2023-11-28 at 4 51 39 PM](https://github.com/simularium/nbsv/assets/24981838/c7f6371f-e389-4916-9548-c7a4b17bf0b1)

Icon clicked:
![Screenshot 2023-11-28 at 4 52 10 PM](https://github.com/simularium/nbsv/assets/24981838/e0c202c3-1df2-4403-9de8-ab3e0428b031)

interim17 committed:
info circle icon

interim17 committed:
viewer title component

interim17 committed:
render viewer title

interim17 committed:
parse traj data and pass down

interim17 committed:
viewer title styling

interim17 committed:
pass down modelInfo

interim17 committed:
check for metadata

interim17 committed:
add font constant

interim17 committed:
adjust margin and font size

interim17 committed:
make background black

interim17 committed:
make background black

interim17 committed:
update meta data check in title

interim17 committed:
remove old comment

interim17 committed:
center the viewer under the title

interim17 committed:
fix viewer center

interim17 committed:
import organize

interim17 committed:
remove widget margins that spill over

interim17 committed:
remove widget margins that spill over

interim17 committed:
remove placeholder controls

interim17 committed:
reposition camera controls

ShrimpCryptid commented on src/components/ViewerTitle.tsx:
```diff
@@ -0,0 +1,36 @@
+  return (
+    <div className="title-container">
+      <div className="title">
+        {props.modelInfo.title}
```
Should there be error handling here for if `props.modelInfo.title` is undefined? Since it looks like it's an optional property in `ModelInfo`.

ShrimpCryptid replied:
Maybe a better question is whether the component should still render, and if it should show the additional metadata in the info button, if there's no defined title.

ShrimpCryptid replied:
![image](https://github.com/simularium/nbsv/assets/30200665/2442c447-2c22-4d78-a0bc-2a4a998de0e9)

Here's what it looks like if no title is defined, also:

interim17 replied:
In theory you could have a trajectory with no title, but with some other meta data, like author names. That case wasn't specified in the designs. My assumption is that we would still want to render the component to make the info panel accessible.

So yes, we could add error handling, maybe if we wanted to adjust the placement of the icon in a case where there is no title provided, but also the default view in your screenshot is an option. I will ask @TravisKroeker. What do you think @ShrimpCryptid, does it look weird having the Icon dead center with no title?

ShrimpCryptid replied:
I kind of like having a placeholder, like "*Untitled Model*" or trajectory or something?

interim17 replied:
Good idea! Sorry lost this suggestion in the shuffle and Megan just reminded me.

ShrimpCryptid commented on src/components/ViewerTitle.tsx:
```diff
@@ -0,0 +1,36 @@
+    <div className="title-container">
+      <div className="title">
+        {props.modelInfo.title}
+        {hasMetaData() ? <div className="info-button"> {Info} </div> : null}
```
Nit: I know from the description that you're going to implement this later, but it might be nice to add a TODO for that in the code itself

interim17 replied:
Done, thanks

ShrimpCryptid commented on css/viewer_title.css:
```diff
@@ -0,0 +1,29 @@
+  justify-content: center;
+}
+
+.title {
+  color: #d8d8d8;
+  display: flex;
+}
+
+.info-button {
+  padding-left: 7px;
+}
+
+.info-button svg {
+  background-color: black;
+  color: white;
+}
+
+.info-button svg:hover {
+  color: #E7E4F2;
+}
+
+.info-button svg:active {
+  color: #B59FF6;
+}
```
Would be nice to define these colors in `constants.css`!

interim17 replied:
Yes, there is an open [PR](https://github.com/simularium/nbsv/pull/8) that defines `colors.css`, just waiting on eyes from Megan and we'll get those converted.

ShrimpCryptid reviewed (approved):
Overall looks sensible! I left a few comments about an edge case with no title defined and moving variables, but none of them are blockers.

ShrimpCryptid commented on css/viewer_title.css:
```diff
@@ -0,0 +1,29 @@
+.title-container {
+  font-size: 20px;
+  font-weight: 400;
+  margin-top: 20px;
+  display: flex;
+  justify-content: center;
+}
+
+.title {
+  color: #d8d8d8;
+  display: flex;
+}
+
+.info-button {
+  padding-left: 7px;
+}
```
Could replace `padding-left` in `.info-button` by adding a `gap: 7px` attribute to `.title-container` instead

interim17 replied:
I like it, less lines

interim17 replied:
Something about merging main broke this, going back to `padding-left` for now.

ShrimpCryptid commented:
Also, agreed that the hover color is really light

interim17 committed:
add TODO comment

interim17 committed:
padding left change to gap

interim17 committed:
move agent colors to constants

interim17 committed:
stop parsing model string

interim17 committed:
handle traj info with viewer methods

interim17 committed:
add colorsheet and use variables (#8)

* add colorsheet and use variables

* fix color names

* rename gray-one

* !important on button hover state

* add button ids and styling

interim17 commented on src/Viewer.tsx:
```diff
@@ -17,58 +21,44 @@ export interface WidgetProps {
-  '#9f516c',
-  '#00aabf',
-];
-
```
Moved agent colors to constants for readability

interim17 commented on src/Viewer.tsx:
```diff
@@ -17,58 +21,44 @@ export interface WidgetProps {
-];
-
 function ViewerWidget(props: WidgetProps): JSX.Element {
+  const [modelInfo, setModelInfo] = useState<ModelInfo>({});
+
+  const handleTrajectoryData = (data: any) => {
+    setModelInfo(data.modelInfo);
+  };
+
```
This is a better way of getting the `modelInfo`, we don't parse the string a second time in `widget.ts`

interim17 committed:
use color variables

interim17 committed:
Merge branch 'main' into feature/viewer-title

meganrm commented:
I haven't seen the title successfully update when it's changed. do you?
<img width="950" alt="Screenshot 2023-11-30 at 3 37 33 PM" src="https://github.com/simularium/nbsv/assets/5170636/abca8fe5-008b-4edd-bfe2-adcced9bab3b">

interim17 committed:
padding-left instead of gap

interim17 commented:
> I haven't seen the title successfully update when it's changed. do you? <img alt="Screenshot 2023-11-30 at 3 37 33 PM" width="950" src="https://private-user-images.githubusercontent.com/5170636/287099838-abca8fe5-008b-4edd-bfe2-adcced9bab3b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDEzODg0NDIsIm5iZiI6MTcwMTM4ODE0MiwicGF0aCI6Ii81MTcwNjM2LzI4NzA5OTgzOC1hYmNhOGZlNS0wMDhiLTRlZGQtYmZlMi1hZGNjZWQ5YmFiM2IucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTEzMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzExMzBUMjM0OTAyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDQ5ZDQxMmQ3NzQ1OTBlMzk4YWZmZTk1NzZmOWYwYzJkMmFjYTgwMzFkYmRmMWU0MGNkYzFiOGVjYTAzOGMyOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.S2Oj8jzGgnFN7n6imUA4ooyu_3ky9Sk_jaWerMbEzXU">

No I'm not seeing it update either.

@meganrm If you notice we're now getting the `modelInfo` with a different method as you requested yesterday, not parsing the `trajectory_str` a second time, but I just tried this with the old method and neither way do we get a new value when running this cell. Am I right in thinking that this is an issue with `update_meta_data`?

interim17 committed:
swap meta data title update argument

interim17 commented:
![Screenshot 2023-11-30 at 4 32 49 PM](https://github.com/simularium/nbsv/assets/24981838/44a75769-aaf5-46af-b585-3d609ef8d3c1)


@meganrm Ok so there are two places we can update something called title. One lives at the`trajectoryInfo` level and one lives under `modelInfo` with the other metadata. To update that run we run this: `w.update_meta_data(title="The coolest model ever!!!!")` instead of this `update_meta_data(trajectory_title="The coolest model ever!!!!")`

I'm pushing a change to the Jupyter cell because I think it makes more sense to change the one that is stored with the other metadata

interim17 committed:
change call of update_meta_data

meganrm commented on src/widget.ts:
```diff
@@ -72,6 +72,7 @@ export class Viewport extends DOMWidgetView {
     const width = this.model.get('width');
     const height = this.model.get('height');
     console.log('rendering viewport', width);
```
probably remove this log

meganrm commented on src/Viewer.tsx:
```diff
@@ -17,58 +21,44 @@ export interface WidgetProps {
-        showPaths={false}
-        onError={console.log}
-      />
+      <ViewerTitle modelInfo={modelInfo} />
```
This spreads out the properties so each one is a prop instead of always having to access `modelInfo.SOMETHING]`
```suggestion
      <ViewerTitle {...modelInfo} />
```

meganrm commented on src/constants.ts:
```diff
@@ -0,0 +1,31 @@
+export interface ModelInfo {
```
this should just be the props for ViewerTitle (which could also be renamed because it's handling all this stuff. Since it's only being used in that file, I would just put it in the ViewerTitle file

meganrm commented on src/components/ViewerTitle.tsx:
```diff
@@ -0,0 +1,37 @@
+  props: ViewerTitleProps
+): JSX.Element => {
+  const hasMetaData = (): boolean => {
+    for (const key in props.modelInfo) {
```
if you make the change I suggested this would become 
```suggestion
    for (const key in props) {
```

meganrm commented on src/components/ViewerTitle.tsx:
```diff
@@ -0,0 +1,37 @@
+  modelInfo: ModelInfo;
+}
+
+const ViewerTitle: React.FunctionComponent<ViewerTitleProps> = (
```
I think a better name for this would be `ModelDisplayData` or something, since it's a lot more than just the title

meganrm commented on src/Viewer.tsx:
```diff
@@ -17,58 +21,44 @@ export interface WidgetProps {
-];
-
 function ViewerWidget(props: WidgetProps): JSX.Element {
+  const [modelInfo, setModelInfo] = useState<ModelInfo>({});
```
oh, I see why you pulled that interface out from here. This seems like something that should exist in the viewer? Instead of having to redefine the interface here. I'll check

meganrm replied:
It was missing, I added it in a few PRs. Since that will take a little bit of time, I would probably just type this as `interface modelInfo {[key]: string}` since it's all optional data anyway, with a TODO to use the TrajectoryFileInfo type exported from the viewer once it's been updated.

interim17 committed:
Update src/Viewer.tsx

Co-authored-by: Megan Riel-Mehan <meganrm@gmail.com>

interim17 committed:
Update src/components/ViewerTitle.tsx

Co-authored-by: Megan Riel-Mehan <meganrm@gmail.com>

meganrm commented on src/Viewer.tsx:
```diff
@@ -17,58 +21,44 @@ export interface WidgetProps {
 function ViewerWidget(props: WidgetProps): JSX.Element {
+  const [modelInfo, setModelInfo] = useState<ModelInfo>({});
+
+  const handleTrajectoryData = (data: any) => {
```
this is where you would type this as `TrajectoryFileInfo` instead of any

interim17 committed:
parse and pass down trajTitle and spread modelInfo

interim17 committed:
receive new props, display either title

interim17 committed:
display meta data even if its just title

meganrm commented on src/components/ViewerTitle.tsx:
```diff
@@ -0,0 +1,37 @@
+  const hasMetaData = (): boolean => {
+    for (const key in props) {
+      if (props[key as keyof ViewerTitleProps] !== undefined) {
+        return true;
```
this needs to still ignore `trajectoryTitle` since it's not part of metadata

interim17 replied:
ack, you're right, fixed

interim17 commented on examples/introduction.ipynb:
```diff
@@ -151,7 +151,7 @@
    ],
    "source": [
-    "w.update_meta_data(trajectory_title=\"The coolest model ever!!!!\")\n",
+    "w.update_meta_data(title=\"The coolest model ever!!!!\")\n",
```
We're now passing down the trajectoryTile (the title for this run), as well as the title from the modelInfo.

To test:
Initial load should show trajectoryTile, i.e. "Some parameter set"
`w.update_meta_data(trajectory_title="The coolest model ever!!!!")` should update the visible title to this text
`w.update_meta_data(trajectory_title="")` should cause display to defer to meta data title, "Some agent baesd model"
`w.update_meta_data(title="new title")` should cause display to show new meta data title, "new title
`w.update_meta_data(trajectory_title="fresh trajTitle)` should reinstate trajectoryTitle and show "fresh trajTitle"

interim17 committed:
exclude trajTitle from metadata check

interim17 committed:
change --light-purple to darker shade

interim17 committed:
add TODO note re: TrajInfoV3 type exports

interim17 committed:
remove redundant console log

interim17 committed:
rename Viewer Title files to ModelDisplayData

interim17 committed:
update names of ModelDisplayData elements

interim17 committed:
update imports for ModelDisplayData

interim17 committed:
update css import

interim17 commented:
Made a few updates:
- added a note to watch for `TrajectoryInfoV3` viewer release
- updated light-purple color after feedback from Travis
- renamed ViewerTitle to ModelDisplayData
- meta data update functions work and title has a priority based display

interim17 committed:
update viewer

interim17 committed:
remove todo comment

interim17 committed:
switch import of modelInfo to use viewer export

interim17 committed:
fix import

interim17 committed:
reorganize import

interim17 committed:
format some imports

interim17 committed:
remove type assertions

meganrm commented on src/components/ModelDisplayData.tsx:
```diff
@@ -0,0 +1,39 @@
+  return (
+    <div className="title-container">
+      <div className="title">
+        {trajectoryTitle || title}
```
I like Peytons idea of having a placeholder title. 
```suggestion
        {trajectoryTitle || title || "<Untitled trajectory>"}
```

interim17 replied:
lol this commit broke the build somehow, i'll follow up shortly

interim17 committed:
Update src/components/ModelDisplayData.tsx

Co-authored-by: Megan Riel-Mehan <meganrm@gmail.com>

ShrimpCryptid commented on src/components/ModelDisplayData.tsx:
```diff
@@ -0,0 +1,39 @@
+  return (
+    <div className="title-container">
+      <div className="title">
+        {trajectoryTitle || title || "<Untitled trajectory>"}
```
```suggestion
        {trajectoryTitle || title || '<Untitled trajectory>'}
```

interim17 replied:
ty

ShrimpCryptid committed:
fix: String formatting

ascibisz reviewed (approved):
looks good to me!
